### PR TITLE
Allow collection expressions with `CreateCompilation`

### PR DESCRIPTION
**Pull request by jcouv**

### CODE REVIEW

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler Small PR for review. Thanks

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler Small PR for review. Thanks

File: src/Compilers/Test/Utilities/CSharp/CSharpTestSource.cs
Comment (jjonescz) [inline src/Compilers/Test/Utilities/CSharp/CSharpTestSource.cs:92]: Can this be static? Or sealed at least.

**Review by jjonescz — APPROVED** [review]

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for second review. Thanks

File: src/Compilers/Test/Core/CollectionBuilderAttribute.cs
Comment (333fred) [inline src/Compilers/Test/Core/CollectionBuilderAttribute.cs:5]: Just to make sure that we don't run afoul of any weird situations (such as razor test dependencies), we should typeforward in an `#else`: https://github.com/dotnet/roslyn/blob/main/docs/Target%20Framework%20Strategy.md#pattern-for-types

File: src/Compilers/Test/Utilities/CSharp/CSharpTestSource.cs
Comment (333fred) [inline src/Compilers/Test/Utilities/CSharp/CSharpTestSource.cs:92]: Why does this need to be a separate type? Why not just make it a static member on this type?

Comment (jcouv) [reply src/Compilers/Test/Utilities/CSharp/CSharpTestSource.cs:92]: It does not *need* to be a separate type, but it is customary for the collection and the collection builder to be different types, and I think it makes sense. See all the collection tests

File: src/Compilers/Test/Core/CollectionBuilderAttribute.cs
Comment (jcouv) [reply src/Compilers/Test/Core/CollectionBuilderAttribute.cs:5]: Thanks for the pointer. That's a useful doc

File: src/Compilers/Test/Utilities/CSharp/CSharpTestSource.cs
Comment (333fred) [reply src/Compilers/Test/Utilities/CSharp/CSharpTestSource.cs:92]: > but it is customary for the collection and the collection builder to be different types

This is a brand new pattern, there's no "customary" at all 😆. I'd just as soon keep it simpler.

Comment (jcouv) [reply src/Compilers/Test/Utilities/CSharp/CSharpTestSource.cs:92]: All our collection builder tests beg to differ. The spec also calls out [two different concepts](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-12.0/collection-expressions.md#create-methods): a "collection type" and a "builder type". It is clearer not to blend them and doesn't seem "simpler".
In any case, this seems a nitpick/preference, not a reason to hold this change up. Thanks

**Review by jaredpar — APPROVED** [review]

**Pull request merged**
